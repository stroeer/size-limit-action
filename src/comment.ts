import { GitHub, context } from '@actions/github';
import { getChangeVars, getFooterVars, change, twoLineBreaks, footer } from './commentTemplate';
import getCommentMarker from './commentMarker';
import { Changes } from './types';

export async function createCommentFromChanges(changes: Changes): Promise<string> {
  return Object.values(changes)
    .map(c => getChangeVars(c))
    .map(v => change(v))
    .concat([footer(getFooterVars(changes))])
    .join(twoLineBreaks);
}

export async function postOrUpdateComment(token: string, comment: string): Promise<void> {
  const octokit = new GitHub(token);
  const reviewId = await getReviewId(octokit);
  if (reviewId) {
    await updateReview(octokit, reviewId, comment);
  } else {
    await postReview(octokit, comment);
  }
}

async function updateReview(octokit: GitHub, review_id: number, body: string): Promise<void> {
  const { owner, repo, number: pull_number } = context.issue;
  await octokit.pulls.updateReview({ owner, repo, pull_number, review_id, body });
}

async function postReview(octokit: GitHub, body: string): Promise<void> {
  const { owner, repo, number: pull_number } = context.issue;
  const event = 'COMMENT';
  await octokit.pulls.createReview({ owner, repo, pull_number, event, body });
}

async function getReviewId(octokit: GitHub): Promise<number | undefined> {
  const { owner, repo, number: pull_number } = context.issue;
  const marker = getCommentMarker();
  const response = await octokit.pulls.listReviews({ owner, repo, pull_number });
  for (const { user, body, id } of response.data) {
    if (user.type === 'Bot' && body.includes(marker)) return id;
  }
}
