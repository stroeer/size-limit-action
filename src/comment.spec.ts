import { createCommentFromChanges } from './comment';

describe('comment', () => {
  test('createCommentFromChanges returns the expected string', async () => {
    const file5 = { bytes: 1, change: -4, path: 'file5.js', percent: -80 };
    const file10 = { bytes: 20, change: 10, path: 'file10.js', percent: 100 };
    const fileX = { bytes: 10, change: 0, path: 'fileX.js', percent: 0 };

    const diff = {
      myPkg: {
        files: [file5, file10, fileX],
        total: { bytes: 31, change: 6, path: 'myPkg', percent: 0.1935483871 },
      },
    };

    await expect(createCommentFromChanges(diff)).resolves.toMatchSnapshot();
  });
});
