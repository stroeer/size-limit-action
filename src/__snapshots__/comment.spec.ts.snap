// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comment createCommentFromChanges returns the expected string 1`] = `
"
## ðŸ“¦ \`myPkg\`

**Size Change:** +6 B (0.19 %)
**Total Size:** 31 B

| Filename           | Size       | Change                      |
| ------------------ | ---------- | --------------------------- |
| \`file5.js\`      | 1 B | -4 B (-80.00 %) |
| \`file10.js\`      | 20 B | +10 B (100.00 %) |

<details>
  <summary><strong>View Unchanged</strong></summary>

  | Filename           | Size       | Change                      |
  | ------------------ | ---------- | --------------------------- |
  | \`fileX.js\`      | 10 B | 0 B (0.00 %) |

</details>

<br />

**Total affected packages:** 1
**Total change:** 6 B
<!-- generated by @paper/size-limit-action -->
"
`;
